### ISSUE TITLE
Wrong code generation + wrong warning

### ISSUE THREAD

**Issue by vporton** [dfinity/motoko#4224]
```motoko
actor {
  let A = 3;

  public query func greet(id : Nat) : async () {
    switch (id) {
      case A {};
      case _ {};
    }
  };
};
```

```
$ moc --version
Motoko compiler 0.10.0+ (source 1w0lk352-sj2y6f0a-b81ma00q-wkf8pimx)
```

In the MS Code UI, the parser gives the warning when the mouse is over `_`:

`this pattern is never matched Motoko(M0146)`.

The warning is wrong. Moreover, I suspect wrong code generation in similar contexts.

**Comment by vporton** [comment]
I confirm wrong code generation:
```motoko
actor {
  let A = 3;

  public query func greet(id : Nat) : async Text {
    switch (id) {
      case A { "error" };
      case _ { "correct" };
    }
  };
};
```
This answers `("error")`, when I input `0`.

**Comment by nomeata** [comment]
I think it works as intended: the `case A` pattern matches a new variable `A` against `id`, shadowing the outer `let A`. It's not possible to perform comparisons with variables in `case`.

**Comment by vporton** [comment]
@nomeata It however should give a warning that it is a match for all possible values (what does not make a useful sense).

**Comment by nomeata** [comment]
Yeah, motoko warnings are not really known for successfully trying to guess what the programmer intended 🤷‍♂️.

But note that it did at least warn you that the second case is unreachable.

If we had warnings about name shadowing, that might have helped here.

**Comment by vporton** [comment]
It is not about name shadowing. I don't want a warning here:
```motoko
let x: ?{y: Nat} = null;
switch (x) {
  case (?x) {};
  case null {};
};
```

I want a warning when a match branch of `switch` is a single variable, because this makes no sense: a single variable match everything and a sane programmer does not want such a match,

**Comment by rossberg** [comment]
> a single variable match everything and a sane programmer does not want such a match,

Of course they do, if it is the last (default) case and they want to do something with the value. If it's not the last case, then you already get a warning that the consecutive case isn't reachable.
